### ISSUE TITLE
Link interconnect_attachment to bgp_peer using attribute reference breaks because of prefix

### ISSUE THREAD

**Issue by skydalriol** [hashicorp/terraform-provider-google#8084]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Terraform Version

v0.12.28

### Affected Resource(s)

resource "google_compute_interconnect_attachment" "vlan_attachment"
resource "google_compute_router_peer" "peer"

### Terraform Configuration Files

```
resource "google_compute_router" "foobar" {
  name    = "router"
  network = google_compute_network.foobar.name
}

resource "google_compute_interconnect_attachment" "on_prem" {
  name         = "on-prem-attachment"
  interconnect = "my-interconnect-id"
  router       = google_compute_router.foobar.id
}

resource "google_compute_router_peer" "peer" {
  name                      = "my-router-peer"
  router                    = google_compute_router.foobar.name
  peer_ip_address           = google_compute_interconnect_attachment.on_prem.customer_router_ip_address
  peer_asn                  = 65500
  advertised_route_priority = 100
  interface                 = "interface-1"
}
```

### Expected Behavior

Given BGP session is strictly linked to the vlan_attachment, I would use the attribute reference of the attachment to obtain the peer IP address (since I don't know beforehand what IP will be assigned to cloud_router_ip_address vs customer_router_ip_address).

### Actual Behavior

```
Error: Error creating RouterBgpPeer: googleapi: Error 400: Invalid value for field 'resource.bgpPeers.peerIpAddress': '169.254.0.2/29'. Must be a valid IPV4 address., invalid
```

The attribute reference from `google_compute_interconnect_attachment` is the IP + Prefix, which cannot be used since `peer_ip_address` only expects the address, without prefix.

### Steps to Reproduce

1. `terraform apply`

### References

https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_interconnect_attachment#cloud_router_ip_address

https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_router_bgp_peer#peer_ip_address

https://github.com/hashicorp/terraform-provider-google/issues/6379

**Comment by skydalriol** [comment]
a workaround I found about getting only the IP address without prefix is the following:

```
resource "google_compute_router_peer" "peer" {
  name                      = "my-router-peer"
  router                    = google_compute_router.foobar.name
  peer_ip_address           = tostring(regex("[^?/]+", google_compute_interconnect_attachment.on_prem.customer_router_ip_address))
  peer_asn                  = 65500
  advertised_route_priority = 100
  interface                 = "interface-1"
}
```

Not ideal but it works :)

**Comment by edwardmedia** [comment]
@skydalriol with or without prefix means different. `google_compute_interconnect_attachment` provides a range of ips, not a single one. You just pick one of IPs, right?  I don't think we should make changes to force the resource to provide one IP instead of a range. What do you think?

**Comment by edwardmedia** [comment]
@skydalriol   closing this as no reply is received. Feel free to reopen it if that does not make sense to you

**Comment by skydalriol** [comment]
@edwardmedia (sorry for re-opening, Xmas break, and btw Happy New Year 😄 )

`You just pick one of IPs` is not entirely correct.
When you create a VLAN attachment, either you specify the `candidate_subnet` or Google will pick one of the available within the range (let it be 169.254.1.0/29 for the sake of this example).

Either way, once the attachment has been created, unless specified during its creation (and unfortunately this cannot be done through Terraform atm), Google automatically assigns one of the IPs to the `cloudRouterIpAddress`, and one to the `customerRouterIpAddress`.

This can be verified by issuing this command:

```gcloud compute interconnects attachments describe on_prem_attachment```

The output (which is correctly returned also as attribute reference on tf) shows which IP is assigned to Google router (169.254.1.1), and which one you have to set on your on-prem router (169.254.1.2) - both within the range (/29 of our example) of the chosen subnet.

This is also explained on the official doc [here](https://cloud.google.com/network-connectivity/docs/interconnect/how-to/dedicated/creating-vlan-attachments#gcloud) (bullet point 2 of the procedure).


So, technically I cannot simply "pick one of the IPs", as that is either assigned by myself during attachment creation, or assigned by Google.

The strange bit is that its output specifies also which range the IP is part of, although entirely useless IMO.
Especially if you consider the next bit (which is about configuring BGP) where you must specify who is who, and for that, you need the single IP address, not the range.

Hope this clarifies my point of view.

**Comment by skydalriol** [comment]
Hi @edwardmedia,

after a bit more testing I understood the reason why the IP+prefix is needed for `cloudRouterIpAddress` and `customerRouterIpAddress` values:

in the first configuration I was missing the `google_compute_router_interface` resource, which is required for the Cloud Router to correctly setup the BGP session, using `google_compute_router_peer` resource.

The interface resource "requires" the ip_range value to be set, otherwise every time a change is made (eg: enabling/disabling the VLAN attachment) this resource gets recreated.

That `ip_range` is actually the `cloud_router_ip_address` value taken from the VLAN attachment, which needs to be specified with the prefix.

So to summarise:

to create a VLAN attachment with Terraform 4 resources are needed:

```
google_compute_router
google_compute_interconnect_attachment
google_compute_router_interface
google_compute_router_peer
```

The router holds the GCP part of the BGP with the advertised ranges.
On the router, an interface needs to be configured.
On the interface you can configure the "on-prem" part of the BGP session.

To link them together, you use the vlan_attachment information taking the `ip_range = cloudRouterIpAddress` for the router_interface configuration, and then you link the "interface_name" into the BGP peer resource configuration.

So to get to my initial point, while at the moment it is not possibile to configure with Terraform the specific IP addresses for CloudRouterIP and CustomerRouterIP, with the `tostring` workaround you can anyway leave to tf the computation of both.

Here the final (working) configuration:

```
resource "google_compute_router" "router" {
  name    = "router"
  network = google_compute_network.foobar.name
 bgp {
    asn               = 64514
 }
}

resource "google_compute_interconnect_attachment" "vlan_attachment" {
  name         = "on-prem-attachment"
  interconnect = "my-interconnect-id"
  router       = google_compute_router.router.id
 candidate_subnets = "169.254.1.0/29"
}

resource "google_compute_router_interface" "int" {
  name       = "interface-1"
  router     = google_compute_router.router.name
  ip_range   = google_compute_interconnect_attachment.vlan_attachment.cloud_router_ip_address
  interconnect_attachment = google_compute_interconnect_attachment.vlan_attachment.name
}

resource "google_compute_router_peer" "peer" {
  name                      = "my-router-peer"
  router                    = google_compute_router_interface.int.name
  peer_ip_address           = tostring(regex("[^?/]+", google_compute_interconnect_attachment.vlan_attachment.customer_router_ip_address))
  interface                 = google_compute_router_interface.int.name
}
```
